Sources/AsyncKit/ConnectionPool/EventLoopGroupConnectionPool.swift

```diff
@@ -178,14 +185,110 @@ public final class EventLoopGroupConnectionPool<Source> where Source: Connection
         }
     }
     
-    deinit {
-        // synchronize access to didShutdown
-        self.lock.lock()
-        defer { self.lock.unlock() }
+    /// Closes the connection pool.
+    ///
+    /// All available connections will be closed immediately. Any connections still in use will be
+    /// closed as soon as they are returned to the pool. Once closed, the pool can not be used to
+    /// create new connections.
+    ///
+    /// Connection pools must be closed before they deinitialize.
+    ///
+    /// This method shuts down synchronously, waiting for all connection closures to complete before
+    /// returning.
+    ///
+    /// - Warning: The pool is always fully shut down once this method returns, even if an error is
+    ///   thrown. All errors are purely advisory.
+    public func syncShutdownGracefully() throws {
+        // - TODO: Does this need to assert "not on any EventLoop", as `EventLoopGroup.syncShutdownGracefully()` does?
+        var possibleError: Error? = nil
+        let waiter = DispatchWorkItem {}
+        let errorLock = Lock()
         
-        if !self.didShutdown {
-            assertionFailure("ConnectionPool.shutdown() was not called before deinit.")
+        self.shutdownGracefully {
+            if let error = $0 {
+                errorLock.withLock { possibleError = error }
+            }
+            waiter.perform()
         }
+        waiter.wait()
+        try errorLock.withLock {
+            if let error = possibleError {
+                throw error
+            }
+        }
+    }
+    
+    /// Closes the connection pool.
+    ///
+    /// All available connections will be closed immediately. Any connections still in use will be
+    /// closed as soon as they are returned to the pool. Once closed, the pool can not be used to
+    /// create new connections.
+    ///
+    /// Connection pools must be closed before they deinitialize.
+    ///
+    /// This method shuts the pool down asynchronously. It may be invoked on any event loop. The
+    /// provided callback will be notified when shutdown is complete. It is invalid to allow a pool
+    /// to deinitialize before it has fully shut down.
+    ///
+    /// This method promises explicitly as API contract not to invoke the callback before returning
+    /// to its caller. It further promises the callback will not be invoked on any event loop
+    /// belonging to the pool.
+    ///
+    /// - Warning: Any invocation of the callback represents a signal that the pool has fully shut
+    ///   down. This is true even if the error parameter is non-`nil`; errors are purely advisory.
+    public func shutdownGracefully(_ callback: @escaping (Error?) -> Void) {
```

Comment:
What is the reason for the callback, versus returning an `EventLoopFuture`?

Reply:
Exactly the same reason `EventLoopGroup` itself does it - to return a future, you need an event loop it's safe to execute on. There is no such loop for the duration of the shutdown (yes, even though it's only shutting down the pool, not the ELG itself); I thought of returning a future created after the shutdown is done (which in theory should be safe) but when I checked with Johannes he pointed out that until I've dispatched to some other thread from whatever called the shutdown method, I can't actually know any of the loops are safe to use - and even if not for that, if the ELG is about to shut down next, which is the most common scenario, forcing a hop to any one of the event loops would end up causing even more context switching than invoking the callback on the dispatch queue does. Pretty much _any_ async shutdown involving an ELG or consumer thereof has to isolate the context which receives the completion notification from the ELG as much as possible,